wrappers/hnynethttp/httptrace.go

```diff
@@ -0,0 +1,200 @@
+			tracer.TLSHandshakeDone(connState, err)
+		},
+		WroteRequest: func(info httptrace.WroteRequestInfo) {
+			tracer.WroteRequest(info)
+		},
+		GotFirstResponseByte: func() {
+			tracer.GotFirstResponseByte()
+		},
+	}
+}
+
+type httpTracer struct {
+	l             sync.Mutex
+	rootSpan      *trace.Span
+	rootCtx       context.Context
+	connectionCtx context.Context
+	dnsCtx        context.Context
+	tlsCtx        context.Context
+	poolCtx       context.Context
+	requestCtx    context.Context
+}
+
+func newTracer(ctx context.Context) *httpTracer {
+	span := trace.GetSpanFromContext(ctx)
+
+	return &httpTracer{rootCtx: ctx, rootSpan: span}
+}
+
+func (h *httpTracer) GetConn(hostPort string) {
+	h.l.Lock()
+	defer h.l.Unlock()
+	if h.rootSpan == nil {
+		return
+	}
+	c, span := h.rootSpan.CreateChild(h.rootCtx)
+	span.AddField("name", "connection")
+	span.AddField("host_port", hostPort)
+	h.connectionCtx = c
+}
+
+func (h *httpTracer) GotConn(info httptrace.GotConnInfo) {
+	h.l.Lock()
+	defer h.l.Unlock()
+	if h.rootSpan == nil {
+		return
+	}
+	span := trace.GetSpanFromContext(h.connectionCtx)
+	span.AddField("conn_was_idle", info.WasIdle)
+	span.AddField("conn_reused", info.Reused)
+	span.AddField("conn_idle_dur_ms", info.IdleTime.Milliseconds)
+	span.Send()
+}
+
+func (h *httpTracer) DNSStart(info httptrace.DNSStartInfo) {
+	h.l.Lock()
+	defer h.l.Unlock()
+	if h.rootSpan == nil {
+		return
+	}
+	c, span := h.rootSpan.CreateChild(h.rootCtx)
+	span.AddField("name", "dns")
+	span.AddField("dns_host", info.Host)
+	h.dnsCtx = c
+}
+
+func (h *httpTracer) DNSDone(info httptrace.DNSDoneInfo) {
+	h.l.Lock()
+	defer h.l.Unlock()
+	if h.rootSpan == nil {
+		return
+	}
+	span := trace.GetSpanFromContext(h.dnsCtx)
+	span.AddField("dns_coalesced", info.Coalesced)
+
+	addrs := []string{}
+	for _, addr := range info.Addrs {
+		addrs = append(addrs, addr.IP.String())
+	}
+	strings.Join(addrs, ",")
+	span.AddField("dns_addrs", info.Addrs)
+	span.Send()
+}
+
+func (h *httpTracer) TLSHandshakeStart() {
+	h.l.Lock()
+	defer h.l.Unlock()
+	if h.rootSpan == nil {
+		return
+	}
+	c, span := h.rootSpan.CreateChild(h.rootCtx)
+	span.AddField("name", "tls")
+	h.tlsCtx = c
+}
+func (h *httpTracer) TLSHandshakeDone(connState tls.ConnectionState, err error) {
+	h.l.Lock()
+	defer h.l.Unlock()
+	if h.rootSpan == nil {
+		return
+	}
+	span := trace.GetSpanFromContext(h.tlsCtx)
+
+	span.AddField("tls_version", connState.Version)
+	span.AddField("tls_server_name", connState.ServerName)
+	span.AddField("tls_negotiated_proto_name", connState.NegotiatedProtocol)
+	span.AddField("tls_cipher_suite", connState.CipherSuite)
+	span.Send()
+}
+
+func (h *httpTracer) ConnectStart(network, addr string) {
+	h.l.Lock()
+	defer h.l.Unlock()
+	if h.rootSpan == nil {
+		return
+	}
+	c, span := h.rootSpan.CreateChild(h.rootCtx)
+	span.AddField("name", "connect")
+	span.AddField("network", network)
+	span.AddField("addr", addr)
+	h.connectionCtx = c
```

Comment:
`Connect()` can be called more than once because Go issues parallel ipv4/ipv6 fetches. Thus, there is not _one_ `connectionCtx`, you'll be creating (and orphaning) one span for the connect that happens second.